---
layout: default
title: graph mashup
category: projects
---
#Graph Mashup
I work with epidemiologists, who do a lot of thinking about statistics and visualizing data. Unfortunately, they also spend a lot of time messing around in some kind of word processor or whatever trying to get the output from various tools to align correctly.

In this project, there is output from a variety of sources. 

- There's a graph that shows the probability that certain genetic mutations are related to certain disease markers.
- There's a output from a genome browser that shows genes and known functions of the genes in the same part of the chromosome as the mutations.
- There is a graph that shows the association between different mutation locations. (I am a little fuzzy on the science part of that myself.)

<img src = "/assets/qtl_graph.png" width = "100%"/> 

In order to build an image like this, the researchers needed to run an R script against their data, look up the second and third parts on two different web sites, and then paste everything together and try to match up the axes of the different graphs. Building such a diagram could take the researcher as long as an hour.

Of course, it's a perfect task for automation. My app runs the R script (via a conversation with an Rserver daemon), scrapes the two websites, and pastes everything together programmatically.  

There's not much interesting in the app logic that does the image stuff...open some connections, make sure everything arrived, and build up the composite. But when you deal with genomics data, the size of the data becomes a significant issue. You can't just shoot off an R job that crunches two gigs of data and hope it finished before the user gets bored or the connection times out.

The secret, at least in this case, is pre-processing. My giant data files represent data for a phenotype over the whole genome, but I only need to graph one chromosome at a time. Splitting the files into 23 parts (and giving R plenty of memory) brought the performance where it needed to be. 

